### PR TITLE
@transifex/api: Update specs and add helpers

### PR DESCRIPTION
**Pull request by nbasili**

### CODE REVIEW

**Review by kbairak — DISMISSED** [review]

File: packages/api/src/transifexApi.js [hunk 1]
```diff
@@ -39,15 +90,52 @@ class ResourceString extends JsonApiResource {
...
+   * @param {object} data
+   * @param {Resource} data.resource
+   * @param {string} data.content
+   * @param {string} data.content_encoding - "text" (default) or "base64"
+   * @param {string} data.callback_url - The url that will be called when processing is complete
+   * @param {number} data.interval - seconds to check status (5sec default)
+   * @memberof ResourceStringsAsyncUpload
+   */
+  static async upload(data) {
+    return upload.call(this, { content_encoding: 'text', data });
```
Comment (nbasili) [inline packages/api/src/transifexApi.js:135]: @kbairak Should this be `{ , ...data }`?

File: packages/api/src/transifexApi.js [hunk 2]
```diff
@@ -71,43 +159,157 @@ TransifexApi.register(ContextScreenshot, 'ContextScreenshot');
...
+   * @param {Language} data.language
+   * @param {string} data.content
+   * @param {string} data.content_encoding - "text" (default) or "base64"
+   * @param {string} file_type - "default" (default) or "xliff"
+   * @param {string} data.callback_url - The url that will be called when processing is complete
+   * @param {number} data.interval - seconds to check status (5sec default)
+   * @memberof ResourceTranslationsAsyncUpload
+   */
+  static async upload(data) {
+    return upload.call(this, { content_encoding: 'text', data });
```
Comment (nbasili) [inline packages/api/src/transifexApi.js:309]: @kbairak Should this be `{ , ...data }`?

File: packages/api/src/transifexApi.js [hunk 3]
```diff
@@ -121,12 +323,54 @@ TransifexApi.register(Team, 'Team');
...
+   * @param {object} data
+   * @param {Project} data.project
+   * @param {Language} data.language
+   * @param {string} data.callback_url - The url that will be called when processing is complete
+   * @param {number} data.interval - seconds to check status (5sec default)
+   * @memberof TmxAsyncDownload
+   * @returns -an axios response
+   */
+  static async download(data) {
+    return download.call(this, { callback_url: null, data });
```
Comment (nbasili) [inline packages/api/src/transifexApi.js:340]: @kbairak Should this be `{ , ...data }`?

File: packages/api/src/transifexApi.js [hunk 4]
```diff
@@ -121,12 +323,54 @@ TransifexApi.register(Team, 'Team');
...
+   * @param {Language} data.language
+   * @param {string} data.content
+   * @param {string} data.content_encoding - "text" (default) or "base64"
+   * @param {boolean} data.override - Whether to override the TM
+   * @param {string} data.callback_url - The url that will be called when processing is complete
+   * @param {number} data.interval - seconds to check status (5sec default)
+   * @memberof TmxAsyncUpload
+   */
+  static async upload(data) {
+    return upload.call(this, { content_encoding: 'text', data });
```
Comment (nbasili) [inline packages/api/src/transifexApi.js:366]: Should this be `{ , ...data }`?

File: packages/api/src/transifexApi.js [hunk 5]
```diff
@@ -1,4 +1,33 @@
...
+      resolve();
+    }, sec * 1000);
+  });
+}
+
+async function download({ interval = 5, ...data }) {
+  // 'this' is a Resource subclass
+  const resource = await this.create({ content_encoding: 'text', ...data });
+  while (!resource.redirect) {
+    /* eslint-disable no-await-in-loop */
```
Comment (nbasili) [inline packages/api/src/transifexApi.js:16]: I do not think this is needed, as it is disabled for the entire file.

File: packages/api/src/transifexApi.js [hunk 6]
```diff
@@ -1,4 +1,33 @@
...
+    await sleep(interval);
+    await resource.reload();
+  }
+  return resource.follow();
+}
+
+async function upload({ interval = 5, ...data }) {
+  const resource = await this.create({ content_encoding: 'text', ...data });
+  while (resource.get('status') !== 'succeeded') {
+    /* eslint-disable no-await-in-loop */
```
Comment (nbasili) [inline packages/api/src/transifexApi.js:26]: I do not think this is needed, as it is disabled for the entire file.

File: packages/api/src/transifexApi.js [hunk 7]
```diff
@@ -1,4 +1,33 @@
...
+  return new Promise((resolve) => {
+    setTimeout(() => {
+      resolve();
+    }, sec * 1000);
+  });
+}
+
+async function download({ interval = 5, ...data }) {
+  // 'this' is a Resource subclass
+  const resource = await this.create({ content_encoding: 'text', ...data });
```
Comment (nbasili) [inline packages/api/src/transifexApi.js:14]: We need to decide if we will be passing the `content_encoding` defaults here or in the calling method. 
Maybe setting the defaults there is better to keep it more abstract, like you do in other places of the code:
`upload.call(this, { content_encoding: 'text', ...data })`

File: packages/api/src/transifexApi.js [hunk 8]
```diff
@@ -1,4 +1,33 @@
...
+  while (!resource.redirect) {
+    /* eslint-disable no-await-in-loop */
+    await sleep(interval);
+    await resource.reload();
+  }
+  return resource.follow();
+}
+
+async function upload({ interval = 5, ...data }) {
+  const resource = await this.create({ content_encoding: 'text', ...data });
```
Comment (nbasili) [inline packages/api/src/transifexApi.js:24]: We need to decide if we will be passing the `content_encoding` defaults here or in the calling method. 
Maybe setting the defaults there is better to keep it more abstract, like you do in other places of the code:
`upload.call(this, { content_encoding: 'text', ...data })`

File: packages/api/src/transifexApi.js (hunk 7 above)
Comment (kbairak) [reply packages/api/src/transifexApi.js:14]: Yes, I meant to remove it from the general `download` function and add it as a default in the more specific ones.

File: packages/api/src/transifexApi.js [hunk 9]
```diff
@@ -39,15 +88,52 @@ class ResourceString extends JsonApiResource {
...
+   * @param {string} data.content_encoding - "text" (default) or "base64"
+   * @param {string} data.file_type - "default" (default) or "xliff"
+   * @param {boolean} data.pseudo - Generate mock string translations
+   * @param {string} data.callback_url - The url that will be called when processing is complete
+   * @param {number} data.interval - seconds to check status (5sec default)
+   * @memberof ResourceStringsAsyncDownload
+   * @returns -an axios response
+   */
+  static async download(data) {
+    return download.call(this, { content_encoding: 'text', ...data });
```
Comment (nbasili) [inline packages/api/src/transifexApi.js:109]: Should we also pass the `file_type: 'default'` here as well? Is it needed?

File: packages/api/src/transifexApi.js [hunk 10]
```diff
@@ -71,43 +157,157 @@ TransifexApi.register(ContextScreenshot, 'ContextScreenshot');
...
+   * @static
+   * @param {object} data
+   * @param {Organization} data.organization
+   * @param {Language} data.language
+   * @param {User} data.user
+   * @param {string} data.date_from - Date, will be treated as UTC by the API
+   * @param {string} data.date_to - Date, will be treated as UTC by the API
+   * @param {number} data.interval - seconds to check status (5sec default)
+   * @memberof OrganizationActivityReportsAsyncDownload
+   * @returns -an axios response
```
Comment (nbasili) [inline packages/api/src/transifexApi.js:173]: This should be a URL, right?

File: packages/api/src/transifexApi.js [hunk 11]
```diff
@@ -71,43 +157,157 @@ TransifexApi.register(ContextScreenshot, 'ContextScreenshot');
...
+   * @static
+   * @param {object} data
+   * @param {Project} data.project
+   * @param {Language} data.language
+   * @param {User} data.user
+   * @param {string} data.date_from - Date, will be treated as UTC by the API
+   * @param {string} data.date_to - Date, will be treated as UTC by the API
+   * @param {number} data.interval - seconds to check status (5sec default)
+   * @memberof ProjectActivityReportsAsyncDownload
+   * @returns -an axios response
```
Comment (nbasili) [inline packages/api/src/transifexApi.js:196]: This should be a URL, right?

File: packages/api/src/transifexApi.js [hunk 12]
```diff
@@ -71,43 +157,157 @@ TransifexApi.register(ContextScreenshot, 'ContextScreenshot');
...
+   * @static
+   * @param {object} data
+   * @param {Resource} data.resource
+   * @param {Language} data.language
+   * @param {User} data.user
+   * @param {string} data.date_from - Date, will be treated as UTC by the API
+   * @param {string} data.date_to - Date, will be treated as UTC by the API
+   * @param {number} data.interval - seconds to check status (5sec default)
+   * @memberof ResourceActivityReportsAsyncDownload
+   * @returns -an axios response
```
Comment (nbasili) [inline packages/api/src/transifexApi.js:219]: This should be a URL, right?

File: packages/api/src/transifexApi.js [hunk 13]
```diff
@@ -71,43 +157,157 @@ TransifexApi.register(ContextScreenshot, 'ContextScreenshot');
...
+   * @static
+   * @param {object} data
+   * @param {Team} data.team
+   * @param {Language} data.language
+   * @param {User} data.user
+   * @param {string} data.date_from - Date, will be treated as UTC by the API
+   * @param {string} data.date_to - Date, will be treated as UTC by the API
+   * @param {number} data.interval - seconds to check status (5sec default)
+   * @memberof TeamActivityReportsAsyncDownload
+   * @returns -an axios response
```
Comment (nbasili) [inline packages/api/src/transifexApi.js:242]: This should be a URL, right?

File: packages/api/src/transifexApi.js [hunk 14]
```diff
@@ -71,43 +157,157 @@ TransifexApi.register(ContextScreenshot, 'ContextScreenshot');
...
+   * @param {Language} data.language
+   * @param {string} data.content_encoding - "text" (default) or "base64"
+   * @param {string} data.file_type - "default" (default) or "xliff"
+   * @param {string} data.mode - "default" (default), "reviewed", "proofread",
+   *   "translator", "untranslated", "onlytranslated", "onlyreviewed",
+   *   "onlyproofread", "sourceastranslation"
+   * @param {string} data.callback_url - The url that will be called when processing is complete
+   * @param {number} data.interval - seconds to check status (5sec default)
+   * @memberof ResourceTranslationsAsyncDownload
+   * @returns -an axios response
```
Comment (nbasili) [inline packages/api/src/transifexApi.js:278]: This should be a URL, right?

File: packages/api/src/transifexApi.js [hunk 15]
```diff
@@ -71,43 +157,157 @@ TransifexApi.register(ContextScreenshot, 'ContextScreenshot');
...
+   * @param {Language} data.language
+   * @param {string} data.content
+   * @param {string} data.content_encoding - "text" (default) or "base64"
+   * @param {string} file_type - "default" (default) or "xliff"
+   * @param {string} data.callback_url - The url that will be called when processing is complete
+   * @param {number} data.interval - seconds to check status (5sec default)
+   * @memberof ResourceTranslationsAsyncUpload
+   */
+  static async upload(data) {
+    return upload.call(this, { content_encoding: 'text', ...data });
```
Comment (nbasili) [inline packages/api/src/transifexApi.js:307]: Should we also set the `file_type: 'default'` here?

File: packages/api/src/transifexApi.js [hunk 16]
```diff
@@ -121,12 +321,54 @@ TransifexApi.register(Team, 'Team');
...
+   * Export project's TM as a TMX file
+   *
+   * @static
+   * @param {object} data
+   * @param {Project} data.project
+   * @param {Language} data.language
+   * @param {string} data.callback_url - The url that will be called when processing is complete
+   * @param {number} data.interval - seconds to check status (5sec default)
+   * @memberof TmxAsyncDownload
+   * @returns -an axios response
```
Comment (nbasili) [inline packages/api/src/transifexApi.js:335]: This should be a URL, right?

**Review by kbairak — APPROVED** [review]

**Pull request merged**